File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

// calculate luhn's algorithm as per cs50 pset1 credit spec :)
int main(void)
{
    // prompt user for positive long long int number
    long long int digits;
    do
    {
        digits = get_long_long("Number: ");
    } while (digits < 0);
    // set variables for total digits, and to store products of even and odd numbers
    int digit_count = log10(digits) + 1;
    int digit_product_even = 0;
    int digit_product_odd = 0;
    // iterate backwards through digit int to calculate product of digits
    for (long long int i = digit_count, j = 1; i > 0; i -= 2, j *= 100)
    {
        int dig = digits / (10 * j) % 10;
        digit_product_odd += digits / (1 * j) % 10;
        // digit * 2 on numbers over 4 will return two numbers which need to be added separately
        if (dig * 2 > 9)
        {
            digit_product_even += dig * 2 % 10;
            digit_product_even += dig * 2 / 10;
        }
        else
        {
            digit_product_even += dig * 2;
        }
    }
    // if last digit of combined products is 0
    // then check the digit amount and starting digits to confirm card type
    if ((digit_product_even + digit_product_odd) % 10 == 0)
    {
        if ((digit_count == 16 && digits / 1000000000000000 == 4) || (digit_count == 13 && digits / 1000000000000 == 4))
        {
            printf("VISA\n");
        }
        else if (digit_count == 16 && digits / 100000000000000 > 50 && digits / 100000000000000 < 56)
        {
            printf("MASTERCARD\n");
        }
        else if (digit_count == 15 && digits / 10000000000000 == 37)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: pset6/credit/credit.py
import math


def get_positive_int(prompt):
    """Ask user for a positive integer between 1 and 23"""
    while True:
        try:
            num = int(input("{}: ".format(prompt)), 10)
        except ValueError:
            continue
        if num > 0:
            return num
        else:
            continue


def luhn():
    """Calculate luhn's algorithm as per cs50 pset1 credit spec :) """
    card_number = get_positive_int("Number")
    digit_count = math.ceil(math.log10(card_number))
    card_str = [int(i) for i in str(card_number)]
    first_two_digits = card_str[0] * 10 + card_str[1]
    digit_product_odd = [x for i in range(
        digit_count - 1, 0, -2) for x in list(str(card_str[i-1] * 2))]
    digit_product_even = [str(card_str[i-1])
                          for i in range(digit_count, 0, -2)]
    if sum(int(i) for i in digit_product_odd + digit_product_even):
        if digit_count == 15 and first_two_digits == 37 or first_two_digits == 34:
            print("AMEX")
        elif digit_count == 13 or 15 and card_str[0] == 4:
            print("VISA")
        elif digit_count == 16 and card_str[0] == 5 and first_two_digits >= 51 and first_two_digits <= 55:
            print("MASTERCARD")
        else:
            print("INVALID")
    else:
        print("INVALID")


if __name__ == "__main__":
    luhn()
